### PR TITLE
Implement MatrixFreeHierarchyHelpers

### LINKED ISSUE #91: MatrixFreeHierarchyHelpers
We need to implement  `MatrixFreeHierarchyHelpers` to use the matrix-free implementation the easiest will probably to derive it (for now) frm `DealIIHierarchyHelpers`.

### PR DESCRIPTION
**Pull request by dalg24**
Closes #91 

~~Drop e9116f5 before merging the changes.~~ Done!
~~Probably should define another evaluator type name for it and test both for the legacy and the new matrix free types.~~ Added `mesh_evaluator_type` string argument with default value `"DealIIMeshEvaluator"` instead

### CODE REVIEW

File: source/dealii/dealii_matrix_free_hierarchy_helpers.cc [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+  auto restrictor_matrix =
+      std::make_shared<dealii::TrilinosWrappers::SparseMatrix>();
+  auto global_operator = get_global_operator(mesh_evaluator);
+  auto matrix_free_global_operator =
+      std::dynamic_pointer_cast<DealIIMatrixFreeOperator<VectorType>>(
+          global_operator);
+  // ughhh
+  auto system_sparse_matrix = matrix_free_global_operator->get_matrix();
+  // why do we pass system matrix to amge?
```
Comment (dalg24) [inline source/dealii/dealii_matrix_free_hierarchy_helpers.cc]: Need to discuss this with Bruno

File: source/dealii/dealii_matrix_free_hierarchy_helpers.cc [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        dealii_mesh_evaluator->get_constraints(), *system_matrix);
+
+    this->_global_operator.reset(
+        new DealIIMatrixFreeOperator<VectorType>(system_matrix));
+  }
+
+  return this->_global_operator;
+}
+
+// copy/paste from DealIIHierarchyHelpers::build_restrictor()
```
Comment (dalg24) [inline source/dealii/dealii_matrix_free_hierarchy_helpers.cc]: Don't forget to update that comment.

File: include/mfmg/common/hierarchy.hpp [hunk 3]
```diff
@@ -38,9 +39,11 @@ create_hierarchy_helpers(std::shared_ptr<MeshEvaluator const> evaluator)
   {
     int const dim = evaluator->get_dim();
     if (dim == 2)
-      hierarchy_helpers.reset(new DealIIHierarchyHelpers<2, VectorType>());
+      hierarchy_helpers.reset(
+          new DealIIMatrixFreeHierarchyHelpers<2, VectorType>());
     else if (dim == 3)
-      hierarchy_helpers.reset(new DealIIHierarchyHelpers<3, VectorType>());
+      hierarchy_helpers.reset(
+          new DealIIMatrixFreeHierarchyHelpers<3, VectorType>());
```
Comment (dalg24) [inline include/mfmg/common/hierarchy.hpp]: DO NOT MERGE THESE CHANGES AS IS

File: include/mfmg/common/operator.hpp [hunk 4]
```diff
@@ -23,6 +23,8 @@ class Operator
   using operator_type = Operator<VectorType>;
   using vector_type = VectorType;
 
+  virtual ~Operator() = default;
```
Comment (dalg24) [inline include/mfmg/common/operator.hpp]: Make sure Bruno sees this.  If necessary make a separate PR for it.

**Comment by dalg24** [comment]
I don't see how to do this without creating a new class deriving from `DealIIMeshEvaluator` which sounds a bit silly to me.

**Comment by Rombur** [comment]
We could add an input parameter to `DealIIMeshEvaluator` which could then return `DealIIMeshEvaluator` or `DealIIMatrixFreeMeshEvaluator`. Not the most elegant but we need a way to switch the implementation without having to recompile the code everytime.

**Comment by dalg24** [comment]
> We could add an input parameter to `DealIIMeshEvaluator` which could then return `DealIIMeshEvaluator` or `DealIIMatrixFreeMeshEvaluator`. Not the most elegant but we need a way to switch the implementation without having to recompile the code everytime.

Fine by me

**Comment by dalg24** [comment]
Zoltan unit test in test_hierarchy does not seems to like the matrix free implementation.

**Comment by Rombur** [comment]
> Zoltan unit test in test_hierarchy does not seems to like the matrix free implementation.

Yeah, strange. The unit test is doesn't do anything special it just uses a different partition of the mesh. It is just less regular than the other partitions.

**Comment by Rombur** [comment]
So the reason why the Zoltan test fails is a little bit strange. The second time that we call Zoltan we get a different partitioning than the first time. I tried to split the single test into two but it didn't change anything. It looks like Zoltan uses some variables that are not reset correctly between runs. I looked at deal.II's code but everything looks fine. Let's just run this test for a single case and write a comment about this problem with Zoltan.

File: source/dealii/dealii_matrix_free_hierarchy_helpers.cc (hunk 1 above)
Comment (Rombur) [reply source/dealii/dealii_matrix_free_hierarchy_helpers.cc]: See #97
